pkg/client/cache/reflector.go

```diff
@@ -169,14 +249,18 @@ func (r *Reflector) syncWith(items []runtime.Object) error {
 }
 
 // watchHandler watches w and keeps *resourceVersion up to date.
-func (r *Reflector) watchHandler(w watch.Interface, resourceVersion *string, exitWatch <-chan time.Time) error {
+func (r *Reflector) watchHandler(w watch.Interface, resourceVersion *string, exitWatch <-chan interface{}) error {
 	start := time.Now()
 	eventCount := 0
+
+	// Stopping the watcher should be idempotent and if we return from this function there's no way
+	// we're coming back in with the same watch interface.
+	defer w.Stop()
```

Comment:
This line's change is good :)

But otherwise, I think I pretty strongly prefer just plumbing two channels down to watchHandler. 'exitWatch' can be renamed to resyncChan and we can add a 'stopCh'. Resyncing and stopping are just two different things. Also--go design style--if you're using both mutexes and channels, something is wrong. :)